TFGCasper.thy

```diff
@@ -0,0 +1,79 @@
+theory TFGCasper
+
+imports Main HOL.Real CBCCasper ConsensusSafety ExampleProtocols
+
+begin
+
+(* Section 4.4: Casper the Friendly GHOST *)
+(* Definition 4.23: Blocks *)
+type_synonym block = consensus_value
+
+locale Ghost = Protocol +
+  fixes Genesis :: block
+  and NextBlkgen :: "block \<Rightarrow> block set"
+
+fun (in Ghost) B_i :: "nat \<Rightarrow> block set"
+  where
+    "B_i 0 = {Genesis}"
+  | "B_i n = \<Union> {NextBlkgen b | b. b \<in> B_i (n - 1)}"
+
+definition (in Ghost) B :: "block set"
```

Comment:
You can fix `B` in the GHOSY locale like `M` and `Σ` in the protocol locale.

Reply:
```
fun  B_i :: "(block * (block ⇒ block set)) ⇒ nat ⇒ block set"
  where
    "B_i (g, blk_gen) 0 = {g}"
  | "B_i (g, blk_gen) n = ⋃ {blk_gen b | b. b ∈ B_i (g, blk_gen) (n - 1)}"

locale Ghost = Protocol +
  fixes Genesis :: block
  fixes B :: "block set"
  and NextBlkgen :: "block ⇒ block set"

  assumes B_def : "B = ⋃ (B_i (Genesis, NextBlkGen)` ℕ)"
```

Reply:
ea91df7330b1c871e5312d398e53160033b77b44